File: src/contexto.mjs
// §7 Generador de contexto. Codigo, no agente: mismo capitulo y mismo canon dan
// siempre el mismo paquete. El escritor nunca consulta el canon por su cuenta.
import { readFileSync, existsSync } from 'node:fs';
import { normalizarFecha } from './canon.mjs';

/** Los cuatro bloques recortables, en el orden en que se recortan (§7). */
export const ORDEN_DE_RECORTE = ['memoria_larga', 'cronologia', 'reparto_fondo', 'epoca'];

/** Estimacion de tokens suficiente para decidir recortes. */
export function estimarTokens(texto) {
  return Math.ceil(String(texto).length / 4);
}

function primeraFrase(texto) {
  const corte = String(texto).match(/^.*?[.!?](\s|$)/s);
  return (corte ? corte[0] : String(texto)).trim();
}

function ultimasPalabras(texto, cuantas) {
  const palabras = String(texto).trim().match(/\S+/g) ?? [];
  return palabras.slice(-cuantas).join(' ');
}

function leerTexto(ruta) {
  return existsSync(ruta) ? readFileSync(ruta, 'utf8') : '';
}

/**
 * Construye los nueve bloques de §7 para un capitulo.
 * Devuelve el paquete sin recortar; el recorte lo aplica generarContexto.
 */
function bloques(canon, numero, config) {
  const ficha = canon.ficha(numero);
  if (!ficha) throw new Error(`no hay ficha del capitulo ${numero}`);

  const todosLosPersonajes = canon.personajes();
  const enEscena = new Set(ficha.personajes);

  // Encargo: la ficha entera, siempre.
  const encargo = ficha;

  // Personajes: fichas completas de quien sale.
  const personajes = canon.personajes(ficha.personajes);

  // Reparto de fondo: quien no sale pero se menciona en la sinopsis.
  const reparto_fondo = todosLosPersonajes
    .filter((p) => !enEscena.has(p.id))
    .filter((p) => ficha.sinopsis.toLowerCase().includes(p.nombre.toLowerCase()))
    .map((p) => ({ id: p.id, nombre: p.nombre, linea: primeraFrase(p.motivacion) }));

  // Memoria: resumenes de los capitulos aprobados anteriores.
  const anteriores = canon.resumenes().filter((r) => r.capitulo < numero);
  const corte = numero - config.contexto.ventana_resumenes;
  const memoria_reciente = anteriores.filter((r) => r.capitulo >= corte);
  const memoria_larga = anteriores.filter((r) => r.capitulo < corte)
    .map((r) => ({ capitulo: r.capitulo, resumen: primeraFrase(r.resumen) }));

  // Hilos vivos: abiertos y aun no cerrados.
  const hilos_vivos = canon.hilosVivos();

  // Epoca: datos que casan con las etiquetas de la ficha, verificado primero.
  const etiquetas = new Set(ficha.etiquetas.map((e) => e.toLowerCase()));
  const peso = { verificado: 0, sin_verificar: 1, inventado: 2 };
  const epoca = canon.datos()
    .filter((d) => d.etiquetas.some((e) => etiquetas.has(String(e).toLowerCase())))
    .sort((a, b) => (peso[a.estado] - peso[b.estado]) || a.id.localeCompare(b.id));

  // Cronologia: eventos entre la fecha de la ficha anterior y la de esta.
  const fichaAnterior = canon.fichas().filter((f) => f.numero < numero).at(-1) ?? null;
  const desde = fichaAnterior ? normalizarFecha(fichaAnterior.fecha) : null;
  const hasta = normalizarFecha(ficha.fecha);
  const cronologia = canon.eventos().filter((e) => {
    const f = normalizarFecha(e.fecha);
    if (!f || !hasta) return false;
    return (desde === null || f >= desde) && f <= hasta;
  });

  // Enganche: cola literal del capitulo anterior aprobado.
  let enganche = '';
  if (fichaAnterior) {
    const aprobado = canon.intentoAprobado(fichaAnterior.numero);
    if (aprobado) {
      enganche = ultimasPalabras(leerTexto(aprobado.ruta), config.contexto.palabras_enganche);
    }
  }

  return {
    encargo,
    personajes,
    reparto_fondo,
    memoria_reciente,
    memoria_larga,
    hilos_vivos,
    epoca,
    cronologia,
    enganche,
  };
}

/**
 * Serializa el paquete en el orden de §7. Este es el formato que el escritor
 * espera; la skill formato-paquete-contexto describe el mismo contrato en prosa.
 */
export function serializar(paquete) {
  const b = paquete.bloques;
  const partes = [];

  partes.push(`# Encargo del capitulo ${b.encargo.numero}`);
  partes.push([
    `- Titulo: ${b.encargo.titulo}`,
    `- Acto: ${b.encargo.acto}`,
    `- Fecha: ${b.encargo.fecha}`,
    `- Objetivo: ${b.encargo.objetivo}`,
    `- Palabras objetivo: ${b.encargo.palabras_objetivo}`,
    `- Sinopsis: ${b.encargo.sinopsis}`,
  ].join('\n'));

  partes.push('# Personajes en escena');
  partes.push(b.personajes.map((p) => [
    `## ${p.nombre} (${p.id}, ${p.rol})`,
    `- Voz: ${p.voz}`,
    `- Motivacion: ${p.motivacion}`,
    `- Arco: ${p.arco}`,
    `- Donde esta: ${p.ubicacion}`,
    `- Que sabe: ${p.sabe.length ? p.sabe.join(' | ') : '(sin anotar)'}`,
  ].join('\n')).join('\n\n') || '(ninguno)');

  if (b.reparto_fondo.length) {
    partes.push('# Reparto de fondo');
    partes.push(b.reparto_fondo.map((p) => `- ${p.nombre} (${p.id}): ${p.linea}`).join('\n'));
  }

  if (b.memoria_reciente.length) {
    partes.push('# Memoria reciente');
    partes.push(b.memoria_reciente.map((r) => `## Capitulo ${r.capitulo}\n${r.resumen}`).join('\n\n'));
  }

  if (b.memoria_larga.length) {
    partes.push('# Memoria larga');
    partes.push(b.memoria_larga.map((r) => `- Cap. ${r.capitulo}: ${r.resumen}`).join('\n'));
  }

  if (b.hilos_vivos.length) {
    partes.push('# Hilos vivos');
    partes.push(b.hilos_vivos.map((h) => `- (abierto en cap. ${h.capitulo}) ${h.hilo}`).join('\n'));
  }

  if (b.epoca.length) {
    // En el paquete viaja el estado de cada dato: el escritor necesita saber
    // que es firme y que es relleno.
    partes.push('# Epoca');
    partes.push(b.epoca.map((d) => `- [${d.estado}] (${d.categoria}) ${d.dato} — fuente: ${d.fuente}`)
      .join('\n'));
  }

  if (b.cronologia.length) {
    partes.push('# Cronologia');
    partes.push(b.cronologia.map((e) => `- ${e.fecha} (${e.tipo}) ${e.descripcion}`).join('\n'));
  }

  if (b.enganche) {
    partes.push('# Enganche con el capitulo anterior');
    partes.push(`> ${b.enganche}`);
  }

  return partes.join('\n\n');
}

/**
 * §7: si el paquete se pasa del tope se recorta en orden fijo. El encargo, los
 * personajes y los hilos vivos no se recortan; si aun asi no cabe, el capitulo
 * se marca bloqueado en lugar de escribirse con el contexto mutilado.
 */
export function generarContexto(canon, numero, config) {
  const paquete = { capitulo: numero, bloques: bloques(canon, numero, config), recortes: [] };
  const tope = config.contexto.tope_contexto;

  for (const nombre of ORDEN_DE_RECORTE) {
    while (estimarTokens(serializar(paquete)) > tope && paquete.bloques[nombre].length > 0) {
      paquete.bloques[nombre] = paquete.bloques[nombre].slice(0, -1);
      paquete.recortes.push(nombre);
    }
    if (estimarTokens(serializar(paquete)) <= tope) break;
  }

  paquete.texto = serializar(paquete);
  paquete.tokens = estimarTokens(paquete.texto);
  paquete.cabe = paquete.tokens <= tope;
  paquete.recortes = [...new Set(paquete.recortes)];
  return paquete;
}

File: novela/contexto.py
# §7 Generador de contexto. Codigo, no agente: mismo capitulo y mismo canon dan
# siempre el mismo paquete. El escritor nunca consulta el canon por su cuenta.
import math
import re
from pathlib import Path

from .canon import normalizar_fecha

# Los cuatro bloques recortables, en el orden en que se recortan (§7).
ORDEN_DE_RECORTE = ['memoria_larga', 'cronologia', 'reparto_fondo', 'epoca']


def estimar_tokens(texto):
    """Estimacion de tokens suficiente para decidir recortes."""
    return math.ceil(len(str(texto)) / 4)


def _primera_frase(texto):
    corte = re.match(r'^.*?[.!?](\s|$)', str(texto), re.DOTALL)
    return (corte.group(0) if corte else str(texto)).strip()


def _ultimas_palabras(texto, cuantas):
    palabras = re.findall(r'\S+', str(texto).strip())
    return ' '.join(palabras[-cuantas:] if cuantas else palabras)


def _leer_texto(ruta):
    camino = Path(ruta)
    return camino.read_text(encoding='utf-8') if camino.exists() else ''


def _bloques(canon, numero, config):
    """Construye los nueve bloques de §7 para un capitulo.

    Devuelve el paquete sin recortar; el recorte lo aplica generar_contexto.
    """
    ficha = canon.ficha(numero)
    if not ficha:
        raise ValueError('no hay ficha del capitulo {}'.format(numero))

    todos_los_personajes = canon.personajes()
    en_escena = set(ficha['personajes'])

    # Encargo: la ficha entera, siempre.
    encargo = ficha

    # Personajes: fichas completas de quien sale.
    personajes = canon.personajes(ficha['personajes'])

    # Reparto de fondo: quien no sale pero se menciona en la sinopsis.
    sinopsis = ficha['sinopsis'].lower()
    reparto_fondo = [
        {'id': p['id'], 'nombre': p['nombre'], 'linea': _primera_frase(p['motivacion'])}
        for p in todos_los_personajes
        if p['id'] not in en_escena and p['nombre'].lower() in sinopsis
    ]

    # Memoria: resumenes de los capitulos aprobados anteriores.
    anteriores = [r for r in canon.resumenes() if r['capitulo'] < numero]
    corte = numero - config['contexto']['ventana_resumenes']
    memoria_reciente = [r for r in anteriores if r['capitulo'] >= corte]
    memoria_larga = [
        {'capitulo': r['capitulo'], 'resumen': _primera_frase(r['resumen'])}
        for r in anteriores if r['capitulo'] < corte
    ]

    # Hilos vivos: abiertos y aun no cerrados.
    hilos_vivos = canon.hilos_vivos()

    # Epoca: datos que casan con las etiquetas de la ficha, verificado primero.
    etiquetas = {str(e).lower() for e in ficha['etiquetas']}
    peso = {'verificado': 0, 'sin_verificar': 1, 'inventado': 2}
    epoca = sorted(
        (d for d in canon.datos()
         if any(str(e).lower() in etiquetas for e in d['etiquetas'])),
        key=lambda d: (peso.get(d['estado'], 9), d['id']),
    )

    # Cronologia: eventos entre la fecha de la ficha anterior y la de esta.
    previas = [f for f in canon.fichas() if f['numero'] < numero]
    ficha_anterior = previas[-1] if previas else None
    desde = normalizar_fecha(ficha_anterior['fecha']) if ficha_anterior else None
    hasta = normalizar_fecha(ficha['fecha'])

    cronologia = []
    for e in canon.eventos():
        f = normalizar_fecha(e['fecha'])
        if not f or not hasta:
            continue
        if (desde is None or f >= desde) and f <= hasta:
            cronologia.append(e)

    # Enganche: cola literal del capitulo anterior aprobado.
    enganche = ''
    if ficha_anterior:
        aprobado = canon.intento_aprobado(ficha_anterior['numero'])
        if aprobado:
            enganche = _ultimas_palabras(
                _leer_texto(aprobado['ruta']), config['contexto']['palabras_enganche'])

    return {
        'encargo': encargo,
        'personajes': personajes,
        'reparto_fondo': reparto_fondo,
        'memoria_reciente': memoria_reciente,
        'memoria_larga': memoria_larga,
        'hilos_vivos': hilos_vivos,
        'epoca': epoca,
        'cronologia': cronologia,
        'enganche': enganche,
    }


def serializar(paquete):
    """Serializa el paquete en el orden de §7.

    Este es el formato que el escritor espera; la skill formato-paquete-contexto
    describe el mismo contrato en prosa.
    """
    b = paquete['bloques']
    partes = []

    partes.append('# Encargo del capitulo {}'.format(b['encargo']['numero']))
    partes.append('\n'.join([
        '- Titulo: {}'.format(b['encargo']['titulo']),
        '- Acto: {}'.format(b['encargo']['acto']),
        '- Fecha: {}'.format(b['encargo']['fecha']),
        '- Objetivo: {}'.format(b['encargo']['objetivo']),
        '- Palabras objetivo: {}'.format(b['encargo']['palabras_objetivo']),
        '- Sinopsis: {}'.format(b['encargo']['sinopsis']),
    ]))

    partes.append('# Personajes en escena')
    fichas_personaje = '\n\n'.join('\n'.join([
        '## {} ({}, {})'.format(p['nombre'], p['id'], p['rol']),
        '- Voz: {}'.format(p['voz']),
        '- Motivacion: {}'.format(p['motivacion']),
        '- Arco: {}'.format(p['arco']),
        '- Donde esta: {}'.format(p['ubicacion']),
        '- Que sabe: {}'.format(' | '.join(p['sabe']) if p['sabe'] else '(sin anotar)'),
    ]) for p in b['personajes'])
    partes.append(fichas_personaje or '(ninguno)')

    if b['reparto_fondo']:
        partes.append('# Reparto de fondo')
        partes.append('\n'.join('- {} ({}): {}'.format(p['nombre'], p['id'], p['linea'])
                                for p in b['reparto_fondo']))

    if b['memoria_reciente']:
        partes.append('# Memoria reciente')
        partes.append('\n\n'.join('## Capitulo {}\n{}'.format(r['capitulo'], r['resumen'])
                                  for r in b['memoria_reciente']))

    if b['memoria_larga']:
        partes.append('# Memoria larga')
        partes.append('\n'.join('- Cap. {}: {}'.format(r['capitulo'], r['resumen'])
                                for r in b['memoria_larga']))

    if b['hilos_vivos']:
        partes.append('# Hilos vivos')
        partes.append('\n'.join('- (abierto en cap. {}) {}'.format(h['capitulo'], h['hilo'])
                                for h in b['hilos_vivos']))

    if b['epoca']:
        # En el paquete viaja el estado de cada dato: el escritor necesita saber
        # que es firme y que es relleno.
        partes.append('# Epoca')
        partes.append('\n'.join('- [{}] ({}) {} — fuente: {}'.format(
            d['estado'], d['categoria'], d['dato'], d['fuente']) for d in b['epoca']))

    if b['cronologia']:
        partes.append('# Cronologia')
        partes.append('\n'.join('- {} ({}) {}'.format(e['fecha'], e['tipo'], e['descripcion'])
                                for e in b['cronologia']))

    if b['enganche']:
        partes.append('# Enganche con el capitulo anterior')
        partes.append('> {}'.format(b['enganche']))

    return '\n\n'.join(partes)


def generar_contexto(canon, numero, config):
    """§7: si el paquete se pasa del tope se recorta en orden fijo.

    El encargo, los personajes y los hilos vivos no se recortan; si aun asi no
    cabe, el capitulo se marca bloqueado en lugar de escribirse con el contexto
    mutilado.
    """
    paquete = {'capitulo': numero, 'bloques': _bloques(canon, numero, config), 'recortes': []}
    tope = config['contexto']['tope_contexto']

    for nombre in ORDEN_DE_RECORTE:
        while (estimar_tokens(serializar(paquete)) > tope
               and len(paquete['bloques'][nombre]) > 0):
            paquete['bloques'][nombre] = paquete['bloques'][nombre][:-1]
            paquete['recortes'].append(nombre)
        if estimar_tokens(serializar(paquete)) <= tope:
            break

    paquete['texto'] = serializar(paquete)
    paquete['tokens'] = estimar_tokens(paquete['texto'])
    paquete['cabe'] = paquete['tokens'] <= tope
    paquete['recortes'] = list(dict.fromkeys(paquete['recortes']))
    return paquete
